lib/diagnostics_channel.js

```diff
@@ -0,0 +1,141 @@
+'use strict';
+
+const {
+  ObjectDefineProperty,
+  ObjectSetPrototypeOf,
+  ObjectValues,
+} = primordials;
+
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+  }
+} = require('internal/errors');
+
+// TODO(qard): should there be a C++ publisher interface?
+class ActivePublisher {
+  subscribe(subscription) {
+    if (typeof subscription === 'function') {
+      this._subscribers.push(subscription);
+      return;
+    }
+
+    if (typeof subscription === 'object' &&
+        typeof subscription.onMessage === 'function') {
+      let shouldListen = true;
+
+      if (typeof subscription.shouldListen === 'function') {
+        shouldListen = subscription.shouldListen(this.name);
+      }
+
+      if (shouldListen) {
+        this._subscribers.push(subscription.onMessage.bind(subscription));
+      }
+    }
+  }
+
+  shouldPublish() {
+    return true;
+  }
+
+  publish(data) {
+    for (var i = 0; i < this._subscribers.length; i++) {
+      this._subscribers[i](this.name, data);
+    }
+  }
+}
+
+class Publisher {
+  constructor(name) {
+    this._subscribers = undefined;
+    ObjectDefineProperty(this, 'name', {
+      value: name
+    });
+  }
+
+  subscribe(subscription) {
+    ObjectSetPrototypeOf(this, ActivePublisher.prototype);
+    this._subscribers = [];
+    this.subscribe(subscription);
+    ObjectDefineProperty(this, 'ifActive', {
+      value: this
+    });
+  }
+
+  shouldPublish() {
+    return false;
+  }
+
+  publish() {}
+}
+
+class DiagnosticsChannel {
+  constructor() {
+    this._channels = {};
+    this._subscribers = [];
+  }
+
+  publisher(name) {
+    let channel = this._channels[name];
+    if (channel) return channel;
+
+    if (typeof name !== 'string') {
+      throw new ERR_INVALID_ARG_TYPE('publisher', ['string'], name);
+    }
+
+    channel = new Publisher(name);
+
+    // New channels should check existing subscribers
+    for (const subscriber of this._subscribers) {
```

Comment:
avoid calling out to mutable globals

```suggestion
    for (let i = 0; i < this._subscribers.length; i++) {
      const subscriber = this._subscribers[i];
```